infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -311,18 +329,63 @@ private void applyShadowStream(UUID streamId, long snapshot) {
         Stream<OpaqueEntry> shadowStream = shadowOpaqueStream.streamUpTo(snapshot);
 
         Iterator<OpaqueEntry> iterator = shadowStream.iterator();
+        List<SMREntry> smrEntries = new ArrayList<>();
+
+        // Clear the stream before updates are applied atomically.
+        // This is done for all replicated streams except 2 cases -
+        // 1. Merge-only streams
+        // 2. Streams which did not evidence data on either source or sink
+        // as these streams will get trimmed and 'clear' will be a 'data loss'.
+        if (mergeOnlyStreams.contains(streamId)) {
+            log.debug("Do not clear stream={} (merge stream)", streamId);
+        } else if (!replicatedStreamIds.contains(streamId)) {
+            log.trace("No data was written to stream {} on source or sink." +
+                "  Do not clear.", streamId);
+            return;
+        } else {
+            smrEntries.add(CLEAR_ENTRY);
+        }
+
         while (iterator.hasNext()) {
             OpaqueEntry opaqueEntry = iterator.next();
-            List<SMREntry> smrEntries =  opaqueEntry.getEntries().get(shadowStreamId);
+            smrEntries.addAll(opaqueEntry.getEntries().get(shadowStreamId));
+        }
 
-            try (TxnContext txnContext = logReplicationMetadataManager.getTxnContext()) {
-                updateLog(txnContext, smrEntries, streamId);
-                CorfuStoreMetadata.Timestamp ts = txnContext.commit();
-                log.debug("Applied shadow stream for stream {} on address :: {}", streamId, ts.getSequence());
-            }
+        List<SMREntry> buffer = new ArrayList<>();
+        long bufferSize = 0;
+        int numBatches = 1;
 
-            log.debug("Process entries count={}", smrEntries.size());
+        for (SMREntry smrEntry : smrEntries) {
+            if (bufferSize + smrEntry.getSerializedSize() >
+                logReplicationMetadataManager.getRuntime().getParameters()
```

Comment:
I also have a question about the runtime here. I thought on Standby side discovery service there are mainly 2 runtimes:
* One created for config, lockstore, metadata manager, with 1/2 default maximum cache size
* One created in LogReplicationSinkManager and passed down to StreamSnapshotWriter, with default maximum cache size.

The only difference is the cache size. Now that we have a buffer in snapshot sync, I thought we are able to decrease the cache size in this class  and let them share the same runtime, which IMO could make the code cleaner.

Reply:
The role of the cache and this buffer are completely different.  The cache keeps tables in the client's memory for faster access.  The buffer is to collect the entries until they can be atomically applied.  This buffer also does not reside in the client.

You have a valid point that runtimes can be unified.  @annym , do you remember what issue/s we were facing if using the same runtime?

Reply:
I noticed that those changes related to runtime creations were from #3119 , and seems the issue is about cache size. So if it's acceptable to reduce the snapshot writer's runtime cache size by 50%, the two runtimes could be unified.

Reply:
Let us keep track of it as a separate item, as discussed today.